thunx-model/src/main/java/eu/xenit/contentcloud/thunx/predicates/model/FunctionExpression.java

```diff
@@ -27,12 +28,35 @@ default String toDebugString() {
                 this.getTerms().stream().map(Object::toString).collect(Collectors.joining(", ")));
     }
 
+    @Override
+    default ThunkExpression<T> simplify() {
+        return getOperator()
+                .trySimplify(getResultType(), getTerms().stream().map(ThunkExpression::simplify).collect(Collectors.toList()))
+                .map(thunkExpression -> (ThunkExpression<T>) thunkExpression)
+                .orElse(this);
+    }
+
     @RequiredArgsConstructor
     @AllArgsConstructor
     enum Operator {
         // Comparison operators
         EQUALS("eq", Boolean.class, (FunctionExpressionFactory<Boolean>) Comparison::areEqual,
-                values -> values.distinct().count() <= 1),
+                (FunctionSimplifier<Boolean>) values -> {
+                    var availableValues = values.stream()
```

Comment:
Note that this is a lot of boilerplate that does not need to be repeated for all the comparison operators, almost everything is common code that can be shared across all operators that require *all* values to be present before they can be calculated.

I also just implemented this function because it was implemented before.
When we are using OPA, it should never return comparators that have known terms on both sides.